Optimized property lookup in index population

fickludd opened the pull request:
Fixes performance regression in modifying nodes with at least one label and many properties. This was caused by loading all properties on any change, which was done to find out what indexes that needed updating. With this PR, only properties that are relevant to at least one online index are loaded.

fickludd committed:
Sketch of index population that only loads relevant properties

The new NodeUpdates object initially gets all labels for the node, before and after, and
information about the property values that are added, removed or changed. The forIndexes
method then analyses all index descriptors and finds what properties that need loading to
determine what indexes need updating. Using the PropertyLoader it loads these properties,
 and returns a sequence of IndexEntryUpdates.

For online index update the glue that gives the NodeUpdates a real PropertyLoader is still missing.

fickludd committed:
Fix test failures from

thobe commented on community/kernel/src/main/java/org/neo4j/kernel/api/index/NodeUpdates.java:
```diff
@@ -125,218 +118,176 @@ private NodeUpdates( long nodeId, long[] labelsBefore, long[] labelsAfter )
+        while ( toLoad.hasNext() )
+        {
+            int propertyId = toLoad.next();
+            Object value = propertyLoader.loadProperty( nodeId, propertyId );
```
Ouch. Is there any way that we could load all properties in one call instead? Like how it's done in `PropertyContainerProxyHelper`.

thobe commented on community/kernel/src/main/java/org/neo4j/kernel/api/index/NodeUpdates.java:
```diff
@@ -487,66 +503,32 @@ private static boolean propertyValueEqual( Object a, Object b )
-        }
+    public interface PropertyLoader
+    {
+        Object loadProperty( long nodeId, int propertyId );
```
I'm thinking this method could be made more specialized to its purpose:

`void loadProperties( long nodeId, PrimitiveIntSet keys, PropertyLoadSink sink );`

Where `NodeUpdates` would implement `PropertyLoadSink` which would be defined as:
```
interface PropertyLoadSink {
    void loadedProperty( int key, Object value);
}
```
and implemented as:
```
class NodeUpdates implements PropertyLoadSink {
...
    public void loadedProperty(int key, Object value) {
        put(key, unchanged(value));
    }
...
}
```

I'm also thinking that implementations of `PropertyLoader.loadProperties(...)` would be allowed to modify `keys` so that after the call it only contains the keys that have no value.

This would allow avoiding multiple calls to the store layer for loading properties.

thobe replied:
With such a change to `PropertyLoader`, the `NodeUpdates.loadProperties(...)` method would change to:
```
void loadProperties( PropertyLoader propertyLoader, PrimitiveIntSet additionalPropertiesToLoad ) {
    propertyLoader.loadProperties( nodeId, additionalPropertiesToLoad, this );
    for ( PrimitiveIntIterator remaining = additionalPropertiesToLoad.iterator(); remaining.hasNext(); ) {
        knownProperties.put( remaining.next(), noValue );
    }
}
```

thobe commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/IndexingService.java:
```diff
@@ -433,13 +434,25 @@ private void apply( IndexUpdates updates, IndexUpdateMode updateMode )
+                    {
+                        // TODO: we should not use exceptions for this, because that is slow
+                        return null;
+                    }
```
This catch block should never be reached in this case, so it's not so bad.

What might be worse than the slowness is that if this happens it would be a symptom of something gone quite horribly wrong, so perhaps we shouldn't just silently ignore it like this...

thobe commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/IndexingService.java:
```diff
@@ -433,13 +434,25 @@ private void apply( IndexUpdates updates, IndexUpdateMode updateMode )
+                    {
+                        // TODO: we should not use exceptions for this, because that is slow
+                        return null;
+                    }
```
You can (and should) avoid this catch block by using `storeView.getProperty( nodeId, propertyId ).value( null )` instead of `storeView.getProperty( nodeId, propertyId ).value()`.

But, as stated elsewhere, I think this should change to an interface that allows loading multiple properties at once, at which point I think `storeView` could implement that interface (the replacement for `NodeUpdate.PropertyLoader`) directly, allowing us to avoid allocating a lambda here.

thobe commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/MultipleIndexPopulator.java:
```diff
@@ -533,7 +539,18 @@ public boolean visit( NodeUpdates updates ) throws IndexPopulationFailedKernelEx
         private void add( NodeUpdates updates )
         {
-            forEachPopulation( population -> population.add( updates ) );
+            for ( IndexEntryUpdate indexUpdate : updates.forIndexes( populations.keySet(), NO_UNCHANGED_PROPERTIES ) )
```
What makes us know that there are `NO_UNCHANGED_PROPERTIES` here?
It's one of those things that would be nice to have in a comment, so that you don't have to go digging through the code for the reason.

I'm guessing it is because this is called from `StoreViewNodeStoreScan.process(...)` where we fully load all properties of a node into the `NodeUpdates` already, but it would be nice to have things like this in comments, so that someone reading the code or making changes to it only has to verify this assumption, and not completely figure out the whole reason on their own.

thobe commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/MultipleIndexPopulator.java:
```diff
@@ -311,18 +317,17 @@ public void flipAfterPopulation()
     {
-        return populations.stream().flatMapToInt( this::propertyKeyIds ).distinct().toArray();
+        return populations.keySet().stream().flatMapToInt( this::propertyKeyIds ).distinct().toArray();
     }
```
This method is actually pretty inefficient, it is only ever used together with `labelIds()`, it would be better to just have a single for-loop build both arrays at once, the streams just add overhead.

fickludd replied:
That is true, but this is a place where any solution works because it's called once per node scan.

thobe commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/MultipleIndexPopulator.java:
```diff
@@ -533,7 +539,18 @@ public boolean visit( NodeUpdates updates ) throws IndexPopulationFailedKernelEx
-            forEachPopulation( population -> population.add( updates ) );
+            for ( IndexEntryUpdate indexUpdate : updates.forIndexes( populations.keySet(), NO_UNCHANGED_PROPERTIES ) )
+            {
+                IndexPopulation population = populations.get( indexUpdate.descriptor() );
```
This seems to be the only reason why `populations` changed from a List to Map, I wonder if we couldn't somehow allow passing the `IndexPopulation` objects (opaquely) through the `NodeUpdates.forIndexes(...)` method, so that we can avoid this lookup altogether.

It seems to me that we add quite a bit of overhead to the data structures with very little gain. Although passing the `IndexPopulation` objects through might add code complexity instead. Feels like one of those things that we will not know until we try it out.

thobe commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/state/OnlineIndexUpdates.java:
```diff
@@ -181,18 +173,19 @@ private NodeRecord loadNode( long nodeId )
+//                propertiesFromCommandsForNode( propertyCommands );
+//        propertyLoader.nodeLoadProperties( nodeRecord, propertiesById, receiver );
+//        return receiver;
+//    }
```
Just remove it. If it is not used it servers no purpose.

thobe commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/state/storeview/StoreViewNodeStoreScan.java:
```diff
@@ -107,10 +107,8 @@ public void process( NodeRecord node ) throws FAILURE
-                propertyUpdatesVisitor.visit( updates.build() );
-            }
+            // TODO: only visit if has label change or label + prop change?
+            propertyUpdatesVisitor.visit( updates.build() );
```
if `updates` has _any_ label/property changes at all? We already know that since the node is from the persistent store there are no _changes_. And determining if the changes are actually relevant or not is what the implementation of the visitor already does.

What we could do is a crude pre-filtering, where we keep a single boolean flag that we set to true in the `if ( propertyKeyIdFilter.test( propertyKeyId ) )` block, is that what you mean?

I can't see that hurting, so why not just do that instead of leaving a vague comment in the code that is mostly just going to cause confusion for future readers of this code.

thobe commented on community/kernel/src/main/java/org/neo4j/unsafe/batchinsert/internal/BatchInserterImpl.java:
```diff
@@ -443,47 +443,48 @@ private void repopulateAllIndexes() throws IOException, IndexEntryConflictExcept
             // Do a lookup from which property has changed to a list of indexes worried about that property.
-            for ( int i = 0; i < descriptors.length; i++ )
+            for ( IndexEntryUpdate indexUpdate :
+                    updates.forIndexes( Iterables.asIterable( descriptors ), NO_UNCHANGED_PROPERTIES ) )
```
Again, it would be nice to note _why_ `NO_UNCHANGED_PROPERTIES` is appropriate here.

thobe reviewed (changes requested):
I think we can do slightly better.

I would also really like to see the benchmark results to know that the changes made really improves the things they intend to improve.

thobe commented:
This PR also needs rebasing on top of a non-broken base branch.

fickludd commented:
Thanks for the great review @thobe. I do have benchmarking data to prove this works, will share via other channel. In hindsight I realize that I made the PR directly after benchmark confirmation,  without making the final clean-up pass...

Will iterate this and return!

fickludd committed:
Removed unused code

fickludd committed:
Added explanations for cases where NO-OP PropertyLoader is given

fickludd committed:
Implemented iterator for PrimitiveLongStack

fickludd committed:
Reworked PropertyLoaded to load multiple properties

Now load a set of properties in one property chain sweep, and calls
callback on successfully loaded properties. We also no longer throw
EntityNotFoundException if the node has been deleted, as this is expected
when nodes are deleted, and we can safely perform a early return instead.

fickludd committed:
Added crude prefilter in store view node store scan

fickludd committed:
Removed map lookups by making IndexEntryUpdate generic in index

thobe commented on community/kernel/src/main/java/org/neo4j/kernel/api/index/PropertyLoader.java:
```diff
@@ -0,0 +1,44 @@
+import org.neo4j.collection.primitive.PrimitiveIntSet;
+
+/**
+ * TODO: make doc. This should probably not be in API, unless directly used by index providers.
```
I think we should act on this `TODO` before merging.

thobe commented on community/kernel/src/main/java/org/neo4j/kernel/api/schema_new/LabelSchemaSupplier.java:
```diff
@@ -0,0 +1,48 @@
+{
+    LabelSchemaDescriptor schema();
+
+    class Box<VALUE> implements LabelSchemaSupplier
```
`Box` does not appear to be used anywhere.

thobe reviewed (changes requested):
Looks good in general, but there are a few things that don't look like they were completed:

* There is a `TODO` comment that I think should be addressed.
* There is an unused class that I think should be removed.

fickludd committed:
Fixed compilation failures

fickludd committed:
Moved NodeUpdates out of index provider API

- removed unused LabelSchemaSupplier.Box
- moved PropertyLoader out of index provider API
